### ISSUE TITLE
A11Y and HTML validation tools

### ISSUE THREAD

**Issue by florian-sanders-cc** [CleverCloud/clever-components#480]
# Goals

We need tools to:
* check that the HTML we produce is robust (tags are closed, attributes are closed and not duplicated, etc.),
* check things related to a11y that can be checked automatically (alt attributes, aria usage, etc.).

# Solutions

After asking on the Lit slack, we think we could rely on 3 layers:

- linting and checking in the editor (easy since it relies on the [a11y lit eslint plugin](https://open-wc.org/docs/linting/eslint-plugin-lit-a11y/overview/)),
- automated and exhaustive testing, to be used before committing for instance ? (would rely on [Open WC Chai Axe](https://open-wc.org/docs/testing/chai-a11y-axe/), as explained earlier),
- specific and more thorough testing for complex components, to be used before committing just to make sure nothing ever breaks in there without us noticing it ? (would rely on [Modern Web Test Runner - a11y snapshot command](https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot)).
This last one looks quite specific and heavy because it generates an accessible tree so that you can check it. That's why we would only use it for a few components.

# What's missing / Up for debate

From what I've tested, all these tools mainly focus on a11y and some basic HTML checks (tags must be closed, etc.).
We're still missing some pieces for HTML checking:

- Content models: for instance, buttons are phrasing contents. They only accept phrasing content as descendants. Using a `<div>` inside it should raise an error but it does not in these tools. They only check stuff directly related to a11y. We used to consider this an accessibility issue but this is debatable (they are considering removing html validation checking from WCAG + I fail to see the issues it could actually cause).
- [Need more testing, I'll edit later... :sweat_smile: ]

**Comment by florian-sanders-cc** [comment]
I'll check Lit & Friends Slack and also OpenWC !

Ideally I would have loved to rely on [html-validate](https://html-validate.org/) under the hood because it has decent presets, decent API to override some rules and an active community.

**Comment by florian-sanders-cc** [comment]
Asked on the Lit & Friends Slack and here is what I got:
- Editor linting - https://open-wc.org/docs/linting/eslint-plugin-lit-a11y/overview/
- Unit testing
  - https://open-wc.org/docs/testing/chai-a11y-axe/
  - https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot

I'll have a closer look at these when I get the time to see how they integrate with our tooling and whether we can fine tune the rules it uses or not.

**Comment by hsablonniere** [comment]
Nice research. I really want use to have proper automated stuffs for HTML and a11y.

We used to have a Storybook addon for a11y

* added: 9434f18ca4aa9aceafad4c7d236a9fe7d512bf66
* removed (when we switched to prebuilt storybook): c68a25dc36ff402368ced27e72a65edbec4b3b84

I was slow!!!! and was not that helpful.

**Comment by hsablonniere** [comment]
I think I would like to have something automated for all components/stories. Having e2e tests for some components is a different topic.

**Comment by florian-sanders-cc** [comment]
I agree !

I checked [a11y snapshot](https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot) from modern dev.
From what I understand, it generates the accessibility tree so you can check stuff from there. For instance, you check that a given element does have an accessible name.

This kind of checking is a lot more thorough and you cannot use a generic test, you have to write it yourself. I think it's useful for complex but important components like Dialogs, combobox (select / autocomplete dropdowns), etc..

I'll edit the desc to sum things up but in the end we could have 3 layers available:

- linting and checking in the editor (easy since it relies on the [a11y lit eslint plugin](https://open-wc.org/docs/linting/eslint-plugin-lit-a11y/overview/)),
- automated and exhaustive testing, to be used before committing for instance ? (would rely on [Open Web Chai Axe](https://open-wc.org/docs/testing/chai-a11y-axe/), as explained earlier),
- specific and more thorough testing for complex components, to be used before committing just to make sure nothing ever breaks in there without us noticing it ? (would rely on [Modern Web Test Runner - a11y snapshot command](https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot))
